primitives/src/lib.rs

```diff
@@ -104,3 +104,141 @@ mod prelude {
+use core::{fmt, convert};
+
+#[cfg(all(feature = "hex", feature = "alloc"))]
+use encoding::{Decodable, Decoder};
+#[cfg(all(feature = "hex", feature = "alloc"))]
+use internals::write_err;
+
+/// An error type for errors that can occur during parsing of a `Decodable` type from hex.
+#[cfg(all(feature = "hex", feature = "alloc"))]
+enum ParsePrimitiveError<T: Decodable> {
+    /// Tried to decode an odd length string
+    OddLengthString(hex::error::OddLengthStringError),
+    /// Encountered an invalid hex character
+    InvalidChar(hex::error::InvalidCharError),
+    /// A decode error from `consensus_encoding`
+    Decode(<T::Decoder as Decoder>::Error),
+}
+
+#[cfg(all(feature = "hex", feature = "alloc"))]
+impl<T: Decodable> fmt::Debug for ParsePrimitiveError<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::OddLengthString(odd_err) => write_err!(f, "odd length string"; odd_err),
+            Self::InvalidChar(char_err) => write_err!(f, "invalid character"; char_err),
+            // Decoder error types don't have Debug, so we only provide this generic error
+            Self::Decode(_) => write!(f, "failure decoding hex string into {}", core::any::type_name::<T>()),
+        }
+    }
+}
+
+#[cfg(all(feature = "hex", feature = "alloc"))]
+impl<T: Decodable> fmt::Display for ParsePrimitiveError<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { fmt::Debug::fmt(&self, f) }
+}
+
+#[cfg(all(feature = "hex", feature = "alloc"))]
+impl<T: Decodable> From<hex::DecodeVariableLengthBytesError> for ParsePrimitiveError<T> {
+    fn from(dec_err: hex::DecodeVariableLengthBytesError) -> Self {
+        use hex::DecodeVariableLengthBytesError as D;
+
+        match dec_err {
+            D::InvalidChar(err) => Self::InvalidChar(err),
+            D::OddLengthString(err) => Self::OddLengthString(err),
+        }
+    }
+}
+
+#[cfg(all(feature = "hex", feature = "alloc"))]
+impl<T: Decodable> From<convert::Infallible> for ParsePrimitiveError<T> {
+    fn from(never: convert::Infallible) -> Self { match never {} }
+}
+
+#[cfg(all(feature = "hex", feature = "alloc", feature = "std"))]
+impl<T: Decodable> std::error::Error for ParsePrimitiveError<T> {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::OddLengthString(ref e) => Some(e),
+            Self::InvalidChar(ref e) => Some(e),
+            Self::Decode(_) => None,
+        }
+    }
+}
+
+#[cfg(all(feature = "hex", feature = "alloc"))]
+pub(crate) mod hex_codec {
+    use super::{fmt, Decodable, ParsePrimitiveError};
+
+    use encoding::{Encodable, EncodableByteIter};
+    use hex_unstable::{BytesToHexIter, Case};
+
+    /// Writes an Encodable object to the given formatter in the requested case.
+    #[inline]
+    fn hex_write_tx_with_case<T: Encodable + Decodable>(obj: &HexPrimitive<T>, f: &mut fmt::Formatter, case: Case) -> fmt::Result {
```

Comment:
Sorry for net mentioning it last review; this has `_tx_ in the function name but writes a `HexPrimitive`. Perhaps just drop the `_tx_` bit. Can be done as a follow up since this is private.